train_decoder.py

```diff
@@ -0,0 +1,462 @@
+    )
+
+    train_dataloader = create_dataloader(train_urls, shuffle_train)
+    val_dataloader = create_dataloader(val_urls, shuffle_val_test, with_text=True)
+    test_dataloader = create_dataloader(test_urls, shuffle_val_test, with_text=True, is_low=True)
+    return {
+        "train": train_dataloader,
+        "val": val_dataloader,
+        "test": test_dataloader
+    }
+
+
+def create_decoder(device, decoder_config, unets_config):
+    """Creates a sample decoder"""
+    unets = []
+    for i in range(0, len(unets_config)):
+        unets.append(Unet(
+            **unets_config[i]
+        ))
+        unets[i].to(device=device)
+    
+    decoder = Decoder(
+        unet=tuple(unets),  # Must be tuple because of cast_tuple
+        **decoder_config
+    )
+    decoder.to(device=device)
+
+    return decoder
+
+def generate_samples(decoder, dataloader, epoch, device, step, n=5, text_prepend=""):
+    """
+    Generates n samples from the decoder and uploads them to wandb
+    Consistently uses the first n image embeddings from the dataloader
+    If the dataloader returns an extra text key, that is used as the caption
+    """
+    test_iter = iter(dataloader)
+    images = []
+    with torch.no_grad():
+        decoder.eval()
+        for data in test_iter:
+            if len(data) == 3:
+                img, emb, txt = data
+            else:
+                img, emb = data
+                txt = [""] * emb.shape[0]
+            # If the number of images is larger than n, we can remove all embeddings after the first n to speed up inference
+            if img.shape[0] > n:
+                img = img[:n]
+                emb = emb[:n]
+                txt = txt[:n]
+            img = img.to(device=device, dtype=torch.float)
+            emb = emb.to(device=device, dtype=torch.float)
+            decoder.to(device=device) # Don't ask me why this is neccesary
+            sample = decoder.sample(emb)
+            for original_image, generated_image, text in zip(img, sample, txt):
+                # Make a grid containing the original image and the generated image
+                img_grid = torchvision.utils.make_grid([original_image, generated_image])
+                image = wandb.Image(img_grid, caption=text_prepend+text)
+                images.append(image)
+                
+                if len(images) >= n:
+                    decoder.to(device)  # Again, don't ask me why but something here moves the model to the cpu
+                    decoder.train()
+                    return images
+    decoder.train()  # In case we run out of samples before n. Your dataset must be tiny, but whatever. Edge cases.
+
+def save_trainer(base_path, trainer, epoch, step, validation_losses, local_only=True, latest=False, best=False, checkpoint=False):
+    """
+    Saves the state of the trainer and decoder to wandb.
+    """
+    print("====================================== Saving trainer ======================================")
```

Comment:
haha kind of nit, but can you use the `print_ribbon` function from https://github.com/lucidrains/DALLE2-pytorch/blob/main/dalle2_pytorch/train.py#L100